From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Cory McIntire <cory@cpanel.net>
Date: Tue, 12 Dec 2017 10:08:09 -0600
Subject: [PATCH 2/6] Concurrent logging adjustment to fix setuid Apache

Case 76493: Resolve locking mutex conflicts with Apache modules
that log as different users.  Serial logging can't be used to
create or lock mutexes because subsequent calls as different
users cause file ownership problems.

By using concurrent logging (one file per logged transaction),
and adding a per-username directory path (via the patch), we
get around the locking issue because each user writes to its
own directory.

This is similar to how EAccelerator caches files on disk.
(Kurt Newman)
---
 apache2/msc_logging.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/apache2/msc_logging.c b/apache2/msc_logging.c
index c2274c7..90b6e3c 100644
--- a/apache2/msc_logging.c
+++ b/apache2/msc_logging.c
@@ -229,11 +229,18 @@ static char *construct_auditlog_filename(apr_pool_t *mp, const char *uniqueid) {
     apr_time_exp_t t;
     char tstr[300];
     apr_size_t len;
+    apr_uid_t uid;
+    apr_gid_t gid;
+    char *username;
 
     apr_time_exp_lt(&t, apr_time_now());
 
     apr_strftime(tstr, &len, 299, "/%Y%m%d/%Y%m%d-%H%M/%Y%m%d-%H%M%S", &t);
-    return apr_psprintf(mp, "%s-%s", tstr, uniqueid);
+
+    apr_uid_current(&uid, &gid, mp);
+    apr_uid_name_get(&username, uid, mp);
+
+    return apr_psprintf(mp, "/%s%s-%s", username, tstr, uniqueid);
 }
 
 /**
